File: go/test/team_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/world-cup-qualification-sdk/go"
	"github.com/voxgig-sdk/world-cup-qualification-sdk/go/core"

	vs "github.com/voxgig-sdk/world-cup-qualification-sdk/go/utility/struct"
)

func TestTeamEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Team(nil)
		if ent == nil {
			t.Fatal("expected non-nil TeamEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := teamBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "team." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WORLDCUPQUALIFICATION_TEST_TEAM_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		teamRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.team", setup.data)))
		var teamRef01Data map[string]any
		if len(teamRef01DataRaw) > 0 {
			teamRef01Data = core.ToMapAny(teamRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = teamRef01Data

		// LIST
		teamRef01Ent := client.Team(nil)
		teamRef01Match := map[string]any{
			"competition_id": setup.idmap["competition01"],
		}

		teamRef01ListResult, err := teamRef01Ent.List(teamRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, teamRef01ListOk := teamRef01ListResult.([]any)
		if !teamRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", teamRef01ListResult)
		}

	})
}

func teamBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "team", "TeamTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read team test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse team test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"team01", "team02", "team03", "competition01", "competition02", "competition03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WORLDCUPQUALIFICATION_TEST_TEAM_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WORLDCUPQUALIFICATION_TEST_TEAM_ENTID": idmap,
		"WORLDCUPQUALIFICATION_TEST_LIVE":      "FALSE",
		"WORLDCUPQUALIFICATION_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["WORLDCUPQUALIFICATION_TEST_TEAM_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WORLDCUPQUALIFICATION_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewWorldCupQualificationSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WORLDCUPQUALIFICATION_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WORLDCUPQUALIFICATION_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/TeamEntityTest.php
<?php
declare(strict_types=1);

// Team entity test

require_once __DIR__ . '/../worldcupqualification_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class TeamEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = WorldCupQualificationSDK::test(null, null);
        $ent = $testsdk->Team(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = team_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "team." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set WORLDCUPQUALIFICATION_TEST_TEAM_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $team_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.team")));
        $team_ref01_data = null;
        if (count($team_ref01_data_raw) > 0) {
            $team_ref01_data = Helpers::to_map($team_ref01_data_raw[0][1]);
        }

        // LIST
        $team_ref01_ent = $client->Team(null);
        $team_ref01_match = [
            "competition_id" => $setup["idmap"]["competition01"],
        ];

        [$team_ref01_list_result, $err] = $team_ref01_ent->list($team_ref01_match, null);
        $this->assertNull($err);
        $this->assertIsArray($team_ref01_list_result);

    }
}

function team_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/team/TeamTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = WorldCupQualificationSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["team01", "team02", "team03", "competition01", "competition02", "competition03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("WORLDCUPQUALIFICATION_TEST_TEAM_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "WORLDCUPQUALIFICATION_TEST_TEAM_ENTID" => $idmap,
        "WORLDCUPQUALIFICATION_TEST_LIVE" => "FALSE",
        "WORLDCUPQUALIFICATION_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["WORLDCUPQUALIFICATION_TEST_TEAM_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["WORLDCUPQUALIFICATION_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new WorldCupQualificationSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["WORLDCUPQUALIFICATION_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["WORLDCUPQUALIFICATION_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
